### PR TITLE
Proposal for restructuring things a bit

### PR DESCRIPTION
**Pull request by rmohr**
@andreabolognani tried to restructure things a little bit, to follow more common go practices.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+all: fmt build
+
+fmt:
+	go fmt ...
+
+build:
```
Comment (rmohr) [inline Makefile:6]: Removed the stamp thing. I would not care too much about optimizing via Makefile logic. go usually does caching itself. Best to only use Makefiles as entrypoints.

File: pkg/devices/memballoon.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	MemBalloonModelVirtio
+	MemBalloonModelVirtioNonTransitional
+	MemBalloonModelVirtioTransitional
+)
+
+type Memballoon struct {
+	model MemballoonModel
+}
+
+func (self MemballoonModel) String() string {
```
Comment (rmohr) [inline pkg/devices/memballoon.go:16]: Using `String()` here instead of `ToStr`, since `String()` will be automatically picked up by `fmt.Print()`. Also note that you can also attach methods on types like `MemballoonModel`. I guess in you rexample it is important to have the `int` values as base to allow seamless integration with libvirt? If really the `string` value is the important bit, then you can also consider making `type MemballoonModel string`.

File: capi/capi.go [hunk 3]
```diff
@@ -3,54 +3,54 @@ package main
 import "C"
 
 import (
-    "gitlab.com/virtblocks/virtblocks/devices"
-    "gitlab.com/virtblocks/virtblocks/util"
+	"github.com/virtblocks/virtblocks/pkg/devices"
```
Comment (rmohr) [inline capi/capi.go:6]: According to the offline chat this would now be `github.com`.

File: pkg/devices/memballoon.go (hunk 2 above)
Comment (andreabolognani) [reply pkg/devices/memballoon.go:16]: > Using `String()` here instead of `ToStr`, since `String()` will be automatically picked up by `fmt.Print()`.

Makes sense, I imaged there would be some conventional name for this functionality but forgot to research it :)

> Also note that you can also attach methods on types like `MemballoonModel`.

That's pretty nice!

> I guess in you rexample it is important to have the `int` values as base to allow seamless integration with libvirt? If really the `string` value is the important bit, then you can also consider making `type MemballoonModel string`.

Correct, the use of `int` is for interoperability with C. Perhaps we could be more clever here but it seems acceptable for the time being.

File: Makefile (hunk 1 above)
Comment (andreabolognani) [reply Makefile:6]: > Removed the stamp thing. I would not care too much about optimizing via Makefile logic. go usually does caching itself. Best to only use Makefiles as entrypoints.

The `Makefile` is for integration with libvirt and other projects that'll want to consume the C interface, see `Makefile.*.inc.am` in https://github.com/virtblocks/libvirt/tree/virtblocks/src/golang for an example.

The use of `stamp` was an attempt to automatically trigger recompilation of the Go code when necessary, but I didn't really spend any time on it and in fact was not working properly at all, so I'm okay with ditching it for the time being and coming up with something better later.

File: capi/capi.go (hunk 3 above)
Comment (andreabolognani) [reply capi/capi.go:6]: Of course :)

**Comment by andreabolognani** [comment]
Overall changes look mostly good, but I'd really like to see them happen over separate commits so that following along is easier... I'll add more comments inline. It's my first time reviewing code on GitHub so please bear with me :)

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,14 @@
+all: fmt build
+
+fmt:
+	go fmt ...
```
Comment (andreabolognani) [inline Makefile]: This fails for me:

```
$ make fmt
go fmt ...
/usr/lib/golang/src/go/build/zcgo.go
open /usr/lib/golang/src/go/build/zcgo.go.579441327: permission denied
exit status 2
make: *** [Makefile:4: fmt] Error 1
```

but if I call `go fmt` manually on the single files it works fine... Any idea what could be wrong?

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,14 @@
+all: fmt build
```
Comment (andreabolognani) [inline Makefile]: I don't think `all` should depend on `fmt`: a user building the project just wants to build the project, not reformat the code.

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+fmt:
+	go fmt ...
+
+build:
+	cd capi && go build -buildmode c-archive -o libvirtblocks.a capi.go && cc -c -o capi.o capi.c
```
Comment (andreabolognani) [inline Makefile]: Having all commands on a single line makes it hard to see what's going on, please do something like

```
cd capi && \
go build ...
```

instead.

File: capi/capi.go [hunk 7]
```diff
@@ -3,54 +3,54 @@ package main
...
+	var goMemballoon = memballoons[cMemballoon]
+	goMemballoon.SetModel(devices.MemballoonModel(cModel))
 }
 
 //export devices_memballoon_get_model
 func devices_memballoon_get_model(cMemballoon C.int) C.int {
-    var goMemballoon = memballoons[cMemballoon]
-    return C.int(goMemballoon.GetModel())
+	var goMemballoon = memballoons[cMemballoon]
+	return C.int(goMemballoon.GetModel())
```
Comment (andreabolognani) [inline capi/capi.go:47]: The whitespace changes that result from calling `go fmt` should be their own commit.

File: cmd/virtblocks/main.go [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package virtblocks
+
+import (
+	"fmt"
+	"github.com/virtblocks/virtblocks/pkg/devices"
+	"github.com/virtblocks/virtblocks/pkg/util"
```
Comment (andreabolognani) [inline cmd/virtblocks/main.go:6]: Is this the usual naming convention for Go projects? Eg. `pkg/` for the library part, and `cmd/` for programs that make use of it? Sounds good, just making sure I understood it correctly.

What about the `virtblocks/` in `cmd/virtblocks/`? It seems redundant. Is that also part of the usual naming convention?

File: go.mod [hunk 9]
```diff
@@ -0,0 +1,3 @@
+module github.com/virtblocks/virtblocks
+
+go 1.12
```
Comment (andreabolognani) [inline go.mod:3]: Adding module support should be its own commit.

File: pkg/devices/memballoon.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
+package devices
+
+type MemballoonModel int
+
+const (
+	MemBalloonModelNone MemballoonModel = iota
+	MemBalloonModelVirtio
+	MemBalloonModelVirtioNonTransitional
+	MemBalloonModelVirtioTransitional
```
Comment (andreabolognani) [inline pkg/devices/memballoon.go]: The constants should be named `MemballoonModelSomething` instead of `MemBalloonModelSomething`.

File: capi/capi.go (hunk 3 above)
Comment (rmohr) [reply capi/capi.go:6]: Now a separate commit.

File: capi/capi.go (hunk 7 above)
Comment (rmohr) [reply capi/capi.go:47]: Done.

File: cmd/virtblocks/main.go (hunk 8 above)
Comment (rmohr) [reply cmd/virtblocks/main.go:6]: Normally you have a `cmd` folder and then  directories underneath it with the names of the binaries. That was just the first name I could think about.

File: go.mod (hunk 9 above)
Comment (rmohr) [reply go.mod:3]: Done.

File: cmd/virtblocks/main.go (hunk 8 above)
Comment (rmohr) [reply cmd/virtblocks/main.go:6]: And yes `pkg` and `cmd` are pretty much standrad.

**Comment by rmohr** [comment]
@andreabolognani I think I addresssed all comments. Should be ready for another review round.

File: Makefile (hunk 6 above)
Comment (rmohr) [reply Makefile]: Done

File: Makefile (hunk 5 above)
Comment (rmohr) [reply Makefile]: Made it independent for now. Normally it is very fast and since normally CI enforces it sooner or later anyway, many of us find it convenient to have it tied together (but not all).

File: Makefile (hunk 4 above)
Comment (rmohr) [reply Makefile]: Ups, was a type. Should have been `go fmt ./...` and not `go fmt ...`.

**Comment by andreabolognani** [comment]
> @andreabolognani I think I addresssed all comments. Should be ready for another review round.

Looks good, thanks for chopping everything up nicely! And `make fmt` works now, too :)

**Comment by andreabolognani** [comment]
@rmohr one last question: is there a specific reason why the `capi/` directory has not been moved to `pkg/` along with the others?

**Comment by rmohr** [comment]
> @rmohr one last question: is there a specific reason why the `capi/` directory has not been moved to `pkg/` along with the others?

I was just not sure what `capi` stands for. Therefore I left it untouched. :)

**Comment by andreabolognani** [comment]
> > @rmohr one last question: is there a specific reason why the `capi/` directory has not been moved to `pkg/` along with the others?
> 
> I was just not sure what `capi` stands for. Therefore I left it untouched. :)

"C API" - that's basically C bindings for Virt Blocks.

It seems like it would make sense to have that directory under `pkg/` along the others, just like we have a similar directory under `src/` for the Rust part.
